src/main/java/org/dsa/iot/alarm/AbstractProvider.java

```diff
@@ -0,0 +1,164 @@
+/* THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD
+ * TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN
+ * NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
+ * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
+ * IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+package org.dsa.iot.alarm;
+
+import org.slf4j.*;
+import java.util.*;
+
+/**
+ * Skeletal Alarming.Provider.
+ *
+ * @author Aaron Hansen
+ */
+public abstract class AbstractProvider implements Alarming.Provider {
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constants
+    ///////////////////////////////////////////////////////////////////////////
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractProvider.class);
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Fields
+    ///////////////////////////////////////////////////////////////////////////
+
+    protected AlarmService service;
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constructors
+    ///////////////////////////////////////////////////////////////////////////
+
+    public AbstractProvider() {}
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Methods
+    ///////////////////////////////////////////////////////////////////////////
+
+
+    /**
+     * {@inheritDoc} <p/>
+     * This implementation configures the alarm record and calls saveRecord.
+     */
+    @Override public void acknowledge(UUID uuid, String user) {
+        AlarmRecord rec = getAlarm(uuid);
+        if (user == null) {
+            throw new NullPointerException("User");
+        }
+        if (rec.getAckTime() <= 0) {
+            rec.setAckTime(System.currentTimeMillis());
+            rec.setAckUser(user);
+            saveRecord(rec);
+        }
+    }
+
+    /**
+     * {@inheritDoc} <p/>
+     * This implementation configures the alarm record and calls saveRecord, then
+     * calls doAddNote.
+     */
+    @Override public void addNote(UUID uuid, String user, String note) {
+        AlarmRecord rec = getAlarm(uuid);
+        if (!rec.getHasNotes()) {
+            rec.setHasNotes(true);
```

Comment:
Notes will definitely be a separate table in a relational implementation, it would be very inefficient because notes are not commonly accessed.